### PR TITLE
Sign out

### PR DESCRIPTION
**Pull request by brentschmaltz**
Modified 'state' logic. AuthenticationProperties are passed to RedirecToIdentityProviderNotificaiton.
Established pattern for reducing 'null-refs'
Added and cleaned up tests.

### CODE REVIEW

**Comment by Tratcher** [comment]
@suhasj

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 1]
```diff
@@ -200,29 +217,25 @@ protected override void ApplyResponseChallenge()
...
+            // the user did set 'OpenIdConnectMessage.State, add a parameter to the state
+            else
             {
-                Logger.LogInformation(Resources.OIDCH_0035_RedirectToIdentityProviderNotificationSkipped);
-                return;
+                redirectToIdentityProviderNotification.ProtocolMessage.State = OpenIdConnectAuthenticationDefaults.AuthenticationPropertiesKey + "=" + Options.StateDataFormat.Protect(properties) + "&userstate=" + redirectToIdentityProviderNotification.ProtocolMessage.State;
             }
 
-            var redirectUri = redirectToIdentityProviderNotification.ProtocolMessage.CreateAuthenticationRequestUrl();
+            string redirectUri = redirectToIdentityProviderNotification.ProtocolMessage.CreateAuthenticationRequestUrl();
```
Comment (suhasj) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: any particular reason why it is typed to string and not `var` ?

**Comment by HaoK** [comment]
This pr looks all messed up as your last prs changes seem to be showing up here?  Can you start with a clean branch of dev and merge your actual changes over to make this more readable.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 1 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: nope, missed that.

**Comment by brentschmaltz** [comment]
@HaoK  I am not sure what you are referring to. I did a 'git pull origin dev' in SignOut before pushing.  What specifically is mixed up.

**Comment by HaoK** [comment]
If you look at this pr changes did you actually make all of the changes shown as part of this batch of changes?  If not I think your branch commit history is messed up. Instead of doing rebases perhaps you did merges.

**Comment by brentschmaltz** [comment]
@HaoK That is what is odd, the file changes are correct, but the commits seem to be picking up old stuff.

**Comment by brentschmaltz** [comment]
Created new PR.

**Pull request closed without merging**